File: 13_roman_to_integer/src/main.rs
#[allow(dead_code)]
struct Solution;

impl Solution {
    #[allow(dead_code)]
    pub fn roman_to_int(s: String) -> i32 {
        let mut result = 0;

        let mut deq = std::collections::VecDeque::from_iter(s.chars());

        while let Some(c) = deq.pop_front() {
            result += match c {
                'M' => 1000,
                'D' => 500,
                'C' => {
                    let next = deq.front().cloned();
                    if next == Some('M') {
                        deq.pop_front();
                        900
                    } else if next == Some('D') {
                        deq.pop_front();
                        400
                    } else {
                        100
                    }
                }
                'L' => 50,
                'X' => {
                    let next = deq.front().cloned();
                    if next == Some('C') {
                        deq.pop_front();
                        90
                    } else if next == Some('L') {
                        deq.pop_front();
                        40
                    } else {
                        10
                    }
                }
                'V' => 5,
                'I' => {
                    let next = deq.front().cloned();
                    if next == Some('X') {
                        deq.pop_front();
                        9
                    } else if next == Some('V') {
                        deq.pop_front();
                        4
                    } else {
                        1
                    }
                }
                _ => 0,
            }
        }

        result
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn testcase_1() {
        let s = "III".to_string();
        let expected = 3;
        assert_eq!(Solution::roman_to_int(s), expected);
    }

    #[test]
    fn testcase_2() {
        let s = "LVIII".to_string();
        let expected = 58;
        assert_eq!(Solution::roman_to_int(s), expected);
    }

    #[test]
    fn testcase_3() {
        let s = "MCMXCIV".to_string();
        let expected = 1994;
        assert_eq!(Solution::roman_to_int(s), expected);
    }
}

fn main() {}

File: 13_roman_to_integer/main.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    int romanToInt(string s) {
        int result = 0;

        for (int i = 0; i < s.length(); i++) {
            switch (s[i]) {
                case 'M':
                    result += 1000;
                    break;
                case 'D':
                    result += 500;
                    break;
                case 'C':
                    if (i < s.length() - 1 && (s[i + 1] == 'M' || s[i + 1] == 'D')) {
                        result -= 100;
                    } else {
                        result += 100;
                    }
                    break;
                case 'L':
                    result += 50;
                    break;
                case 'X':
                    if (i < s.length() - 1 && (s[i + 1] == 'C' || s[i + 1] == 'L')) {
                        result -= 10;
                    } else {
                        result += 10;
                    }
                    break;
                case 'V':
                    result += 5;
                    break;
                case 'I':
                    if (i < s.length() - 1 && (s[i + 1] == 'X' || s[i + 1] == 'V')) {
                        result -= 1;
                    } else {
                        result += 1;
                    }
                    break;
            }
        }

        return result;
    }
};